Library/Homebrew/sbom.rb

```diff
@@ -63,51 +60,128 @@ def self.create(formula, compiler: nil, stdlib: nil)
     new(attributes)
   end
 
-  sig { params(attributes: Hash).void }
-  def initialize(attributes = {})
-    attributes.each { |key, value| instance_variable_set(:"@#{key}", value) }
+  sig { params(formula: Formula).returns(Pathname) }
+  def self.spdxfile(formula)
+    formula.prefix/FILENAME
   end
 
-  sig { returns(T::Boolean) }
-  def valid?
-    data = to_spdx_sbom
+  sig { params(deps: T::Array[Formula]).returns(T::Array[T::Hash[Symbol, String]]) }
+  def self.runtime_deps_hash(deps)
+    deps.map do |dep|
+      {
+        full_name:         dep.full_name,
+        name:              dep.name,
+        version:           dep.version.to_s,
+        revision:          dep.revision,
+        pkg_version:       dep.pkg_version.to_s,
+        declared_directly: true,
+        license:           SPDX.license_expression_to_string(dep.license),
+        bottle:            dep.bottle_hash,
+      }
+    end
+  end
+
+  sig { params(formula: Formula).returns(T::Boolean) }
+  def self.exist?(formula)
+    spdxfile(formula).exist?
+  end
 
-    schema_string, _, status = Utils::Curl.curl_output(SCHEMA)
+  sig { returns(T::Hash[String, String]) }
+  def self.fetch_schema!
+    return @schema if @schema.present?
+
+    url = SCHEMA_URL
+    target = SCHEMA_CACHE_TARGET
+    quieter = target.exist? && !target.empty?
+
+    curl_args = Utils::Curl.curl_args(retries: 0)
+    curl_args += ["--silent", "--time-cond", target.to_s] if quieter
+
+    begin
+      unless quieter
+        oh1 "Fetching SBOM schema"
+        ohai "Downloading #{url}"
+      end
+      Utils::Curl.curl_download(*curl_args, url, to: target, retries: 0)
+      FileUtils.touch(target, mtime: Time.now)
+    rescue ErrorDuringExecution
+      target.unlink if target.exist? && target.empty?
+
+      if target.exist?
+        opoo "SBOM schema update failed, falling back to cached version."
+      else
+        opoo "Failed to fetch SBOM schema, cannot perform SBOM validation!"
+
+        return {}
+      end
+    end
 
-    opoo "Failed to fetch schema!" unless status.success?
+    @schema = begin
+      JSON.parse(target.read, freeze: true)
+    rescue JSON::ParserError
+      target.unlink
+      opoo "Failed to fetch SBOM schema, cached version corrupted, cannot perform SBOM validation!"
+      {}
+    end
+  end
 
-    require "json_schemer"
+  sig { returns(T::Boolean) }
+  def valid?
+    unless require? "json_schemer"
+      error_message = "Cannot require json_schemer, run `brew install-bundler-gems`!"
+      ENV["HOMEBREW_ENFORCE_SBOM"] ? odie(error_message) : opoo(error_message)
```

Comment:
@Bo98 Please suggest better logic/messaging here if this is what you're blocked on getting a ✅